File: server/common/protocol.py
import struct
import socket
import logging
from typing import Optional, Tuple

HEADER_LEN = 5  # 4 bytes para length + 1 byte para message_type
MAX_MESSAGE_LEN = 8 * 1024 # 8kb

class Protocol:
    @staticmethod
    def send_message(sock: socket.socket, message_type: int, message_bytes: bytes) -> None:
        try:
            length = len(message_bytes)
            if length > MAX_MESSAGE_LEN:
                raise ValueError(f"message too large: {length} bytes")

            # 4 bytes de len + 1 byte del message type
            header = length.to_bytes(4, byteorder="big", signed=False) + bytes([message_type])
            full_message = header + message_bytes

            Protocol._send_exact(sock, full_message)

        except Exception as e:
            logging.error(f"action: send_message | result: fail | error: {e}")
            raise

    @staticmethod
    def receive_message(sock: socket.socket) -> Optional[Tuple[int, bytes]]:
        try:
            header = Protocol._receive_exact(sock, HEADER_LEN)
            if not header:
                return None

            message_length = int.from_bytes(header[:4], byteorder="big", signed=False)
            message_type = header[4]

            if message_length < 0 or message_length > MAX_MESSAGE_LEN:
                logging.error(
                    f"action: receive_message | result: fail | error: invalid_length={message_length}"
                )
                return None

            message_bytes = Protocol._receive_exact(sock, message_length)

            if message_bytes is None:
                return None

            return (message_type, message_bytes)

        except Exception as e:
            logging.error(f"action: receive_message | result: fail | error: {e}")
            return None

    @staticmethod
    def _receive_exact(sock: socket.socket, num_bytes: int) -> Optional[bytes]:
        data = b""
        while len(data) < num_bytes:
            chunk = sock.recv(num_bytes - len(data))
            if not chunk:
                return None
            data += chunk
        return data

    @staticmethod
    def _send_exact(sock: socket.socket, full_message: bytes) -> None:
        total_len = len(full_message)
        bytes_sent = 0
        while bytes_sent < total_len:
            sent = sock.send(full_message[bytes_sent:])
            if sent == 0:
                raise ConnectionError("Socket connection broken")
            bytes_sent += sent

            logging.debug(
                f"action: send_message | result: success | bytes_sent: {bytes_sent}"
            )

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"

	"github.com/op/go-logging"
)

var protocolLog = logging.MustGetLogger("protocol")

type Protocol struct{}

func NewProtocol() *Protocol {
	return &Protocol{}
}

const HEADER_LEN = 4

func (p *Protocol) SendMessage(conn net.Conn, messageBytes []byte) error {
	header := make([]byte, HEADER_LEN)
	binary.BigEndian.PutUint32(header, uint32(len(messageBytes)))

	fullMessage := append(header, messageBytes...)

	return p.sendExact(conn, fullMessage)
}

func (p *Protocol) ReceiveMessage(conn net.Conn) ([]byte, error) {
	header, err := p.receiveExact(conn, HEADER_LEN)
	if err != nil {
		protocolLog.Errorf("action: receive_header | result: fail | error: %v", err)
		return nil, fmt.Errorf("failed to receive header: %w", err)
	}

	messageLength := binary.BigEndian.Uint32(header)

	messageBytes, err := p.receiveExact(conn, int(messageLength))
	if err != nil {
		protocolLog.Errorf("action: receive_message | result: fail | error: %v", err)
		return nil, fmt.Errorf("failed to receive message: %w", err)
	}

	protocolLog.Debugf("action: receive_message | result: success | message_length: %d", messageLength)
	return messageBytes, nil
}

func (p *Protocol) receiveExact(conn net.Conn, numBytes int) ([]byte, error) {
	data := make([]byte, numBytes)
	bytesRead := 0

	for bytesRead < numBytes {
		n, err := conn.Read(data[bytesRead:])
		if err != nil {
			if err == io.EOF {
				return nil, fmt.Errorf("connection closed")
			}
			return nil, err
		}
		if n == 0 {
			return nil, fmt.Errorf("connection closed")
		}
		bytesRead += n
	}

	return data, nil
}

func (p *Protocol) sendExact(conn net.Conn, fullMessage []byte) error {
	bytesSent := 0
	for bytesSent < len(fullMessage) {
		n, err := conn.Write(fullMessage[bytesSent:])
		if err != nil {
			protocolLog.Errorf("action: send_message | result: fail | error: %v", err)
			return fmt.Errorf("failed to send message: %w", err)
		}
		if n == 0 {
			return fmt.Errorf("connection broken")
		}
		bytesSent += n
	}
	protocolLog.Debugf("action: send_message | result: success | bytes_sent: %d", bytesSent)
	return nil
}
